src/layers/tile-layer-utils/TileCache.js

```diff
@@ -126,4 +147,72 @@ export default class TileCache {
   _getTileId(x, y, z) {
     return `${z}-${x}-${y}`;
   }
+
+  _expandIndices(tileIndices){
+    var [minX, minY, maxY, maxX] = [Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY, Number.NEGATIVE_INFINITY, Number.NEGATIVE_INFINITY]
+    for (var tileIndex in tileIndices) {
+      var tile = tileIndices[tileIndex]
+      if(tile.x > maxX)
+        maxX = tile.x
+      if(tile.y > maxY)
+        maxY = tile.y
+      if(tile.y < minY)
+        minY = tile.y
+      if(tile.x < minX)
+        minX = tile.x
+    }
+    maxX = Math.min(maxX, Math.floor(MAX_HEIGHT / (256 * Math.pow(2, -1 * tileIndices[0].z))))
+    maxY = Math.min(maxY, Math.floor(MAX_WIDTH / (256 * Math.pow(2, -1 * tileIndices[0].z))))
+    var extraIndices = []
+    for (var tileIndex in tileIndices) {
+      var tile = tileIndices[tileIndex]
+      if (tile.x == maxX)
+        extraIndices.push({x: tile.x + 1, y: tile.y, z: tile.z})
+        if (tile.y == maxY)
+          extraIndices.push({x: tile.x + 1, y: tile.y + 1, z: tile.z})
+        else if (tile.y == minY)
+          extraIndices.push({x: tile.x + 1, y: tile.y - 1, z: tile.z})
+      else if (tile.x == minX)
+        extraIndices.push({x: tile.x - 1, y: tile.y, z: tile.z})
+        if (tile.y == maxY)
+          extraIndices.push({x: tile.x - 1, y: tile.y + 1, z: tile.z})
+        else if (tile.y == minY)
+          extraIndices.push({x: tile.x - 1, y: tile.y - 1, z: tile.z})
+      else if (tile.y == maxY)
+        extraIndices.push({x: tile.x, y: tile.y + 1, z: tile.z})
+      else if (tile.y == minY)
+        extraIndices.push({x: tile.x, y: tile.y - 1, z: tile.z})
+    }
+    return extraIndices
+  }
+
+  _cacheLayersUp(tileIndices, minZoom) {
+    var extraIndices = []
+    var zoom = tileIndices[0].z
+    var [xIndices, yIndices] = [[], []]
+    tileIndices.forEach((e) => xIndices.push(e.x))
+    tileIndices.forEach((e) => yIndices.push(e.y))
+    var [x, y] = [this._median(xIndices), this._median(yIndices)]
+    while(zoom > minZoom) {
+      zoom = zoom - 1
+      var index = {x: Math.floor(x / Math.pow(2, -1 * zoom)), y: Math.floor(y / Math.pow(2, -1 * zoom)), z: zoom}
```

Comment:
Just thinking about the algebra: 1/a = a^-1... and then multiply the exponents and you get a positive.
```suggestion
      var index = {x: Math.floor(x * Math.pow(2, zoom)), y: Math.floor(y * Math.pow(2, zoom)), z: zoom};
```
or pull out a constant:
```suggestion
      const expansion = Math.pow(2, zoom);
      var index = {x: Math.floor(x * expansion), y: Math.floor(y * expansion), z: zoom};
```